src/main/scala/dpla/ingestion3/executors/MappingExecutor.scala

```diff
@@ -76,91 +79,120 @@ trait MappingExecutor extends Serializable {
+    import spark.implicits._
+
+    // these three Encoders allow us to tell Spark/Catalyst how to encode our data in a DataSet.
+    val oreAggregationEncoder: ExpressionEncoder[Row] = RowEncoder(model.sparkSchema)
+
+    val sc = spark.sparkContext
+
     val successResults: Dataset[Row] = mappingResults
       .filter(tuple => Option(tuple._1).isDefined)
       .map(tuple => tuple._1)(oreAggregationEncoder)
 
-    val failures:  Array[String] = mappingResults
+    val exceptions: Array[String] = mappingResults
       .filter(tuple => Option(tuple._2).isDefined)
       .map(tuple => tuple._2).collect()
 
-    // Begin new error and message handling
-    val messages = MessageProcessor.getAllMessages(successResults)
-
-    val messagesExploded = messages
-      .withColumn("level", explode($"level"))
-      .withColumn("message", explode($"message"))
-      .withColumn("field", explode($"field"))
-      .withColumn("value", explode($"value"))
-      .withColumn("id", explode($"id"))
-      .distinct()
-    // Calling distinct here because I was ending up with qaudruplication of all messages.
-    // I suspect `explode` but the dataset might be recomputed unnecessarily
+    val messages = MessageProcessor.getAllMessages(successResults)(spark)
 
-    val warnings = MessageProcessor.getWarnings(messagesExploded)
-    val errors = MessageProcessor.getErrors(messagesExploded)
+    val warnings = MessageProcessor.getWarnings(messages)
+    val errors = MessageProcessor.getErrors(messages)
 
     // get counts
     val attemptedCount = mappingResults.count() // successResults.count()
     val validCount = successResults.select("dplaUri").where("size(messages) == 0").count()
     val warnCount = warnings.count()
-    val errorCount = errors.count()
+    val errorCount = errors.distinct().count()
 
     val recordErrorCount = MessageProcessor.getDistinctIdCount(errors)
     val recordWarnCount = MessageProcessor.getDistinctIdCount(warnings)
 
     val errorMsgDets = MessageProcessor.getMessageFieldSummary(errors).mkString("\n")
     val warnMsgDets = MessageProcessor.getMessageFieldSummary(warnings).mkString("\n")
 
-    val timeInMs = System.currentTimeMillis()
-    val dateTimeStr = Utils.formatDateTime(timeInMs)
-
-    logger.info(s"Message summary")
-    val mappingSummary = MappingSummaryData(
-      shortName,
-      dateTimeStr,
-      attemptedCount,
-      validCount,
-      warnCount,
-      errorCount,
-      recordWarnCount,
-      recordErrorCount,
-      errorMsgDets,
-      warnMsgDets
-    )
-
-    logger.info(MappingSummary.getSummary(mappingSummary))
-    logger.info(s"Number of exceptions ${failures.length}")
-
-    // Write warn and error messages to CSV files. These should all share the same timestamp, minor work TBD
     val baseLogDir = s"$dataOut/../logs/"
-    val time = timeInMs.toString
 
-    val exceptions = sc.parallelize(failures).toDS()
+    val exceptionsDS = sc.parallelize(exceptions).toDS()
```

Comment:
Do you need to make this into a `DataSet`?  Would it be easier to work with it as a plain `RDD`?  Not a big deal, just curious.